### ISSUE TITLE
timm models accuracy regress due to triton pin bump

### ISSUE THREAD

**Issue by shunting314** [pytorch/pytorch#123126]
### 🐛 Describe the bug

When bumping triton from old pin (989adb9a29496c22a36ef82ca69cad5dad536b9c), to new pin (45fff310c891f5a92d55445adf8cc9d29df5841e) here https://github.com/pytorch/pytorch/pull/123077 . We found there are accuracy test regression.

Command
```
time python benchmarks/dynamo/timm_models.py --backend inductor --amp --accuracy --only mixer_b16_224 --training
```

It pass with the old pin but fail with the new pin.

Update: The problem can also be reproed using levit_128 model.

cc  @htyu @manman-ren

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by manman-ren** [comment]
I will take a look. In general, should we also file a github issue at triton/issues? Or that is redundant.

**Comment by shunting314** [comment]
>  should we also file a github issue at triton/issues? Or that is redundant.

Thanks @manman-ren . My understanding is, we can do that if we need help from OAI triton team.

**Comment by manman-ren** [comment]
I was able to reproduce this
```
E0401 16:49:46.154000 139907086283904 torch/_dynamo/utils.py:1405] RMSE (res-fp64): 0.00146, (ref-fp64): 0.00015 and shape=torch.Size([384]). res.dtype: torch.float32, multiplier: 3.000000, tol: 0.010000
E0401 16:49:46.155000 139907086283904 torch/_dynamo/utils.py:1319] Accuracy failed for key name blocks.2.mlp_tokens.fc1.bias
fail_accuracy
```

Applying the pin update patch and using "make triton" doesn't work. I updated my local Triton/llvm to the hash and was able to repro.

There is a llvm update in this pin update: from 4017f04 Jan 30 to 6f44bb7 March 22. It seems that updating the llvm caused this issue. I initially used llvm 4017f04 + new Triton, there was no accuracy issue.

How do we get the generated py file that actually triggered the issue?

**Comment by manman-ren** [comment]
Oh got it. So every generated kernel can be the cause? How do you usually root-cause this? Is there a bisecting tool that overrides the cubin of each kernel? i.e we have two sets of cubins, one for the current pin, the other for the updated pin. We can try to bisect on the two sets by overriding the cubin file with one from the current set.

**Comment by shunting314** [comment]
> Is there a bisecting tool that overrides the cubin of each kernel? i.e we have two sets of cubins, one for the current pin, the other for the updated pin. We can try to bisect on the two sets by overriding the cubin file with one from the current set.

We don't have such a tool yet. This is the first time I see accuracy regression caused by triton.

I think what you said could be a nice way to dig into this issue.

**Comment by manman-ren** [comment]
I am able to bisect it down to one kernel: triton_red_fused__to_copy_native_layer_norm_16.
Attaching the ptx before and after. 
[old-ptx.txt](https://github.com/pytorch/pytorch/files/14844730/old-ptx.txt)
[new-ptx.txt](https://github.com/pytorch/pytorch/files/14844731/new-ptx.txt)

The difference seems to be fma.rn.f32 instead of "mul.f32" and "add.f32".

**Comment by htyu** [comment]
> I am able to bisect it down to one kernel: triton_red_fused__to_copy_native_layer_norm_16. Attaching the ptx before and after. [old-ptx.txt](https://github.com/pytorch/pytorch/files/14844730/old-ptx.txt) [new-ptx.txt](https://github.com/pytorch/pytorch/files/14844731/new-ptx.txt)
> 
> The difference seems to be fma.rn.f32 instead of "mul.f32" and "add.f32".

Great! Yeah, that seems to be the culprint. fma should only be used in fast math mode.

**Comment by shunting314** [comment]
> BTW, curious, did you replace cubins and rerun the test during triaging?

I'm also curious. But I would point triton cache to an alternative path

**Comment by shunting314** [comment]
Nice finding, @manman-ren ! I rerun the command in the summary with FMA manually disabled in triton, and the test can pass now.  How can we properly fix this?

also cc @Chillee who is trying to disabling FMA in inductor.

**Comment by htyu** [comment]
Perhaps a fast-math flag can be added to a Triton kernel and then passed down to LLVM which already accepts fast-math flag on function level.

**Comment by Chillee** [comment]
https://github.com/pytorch/pytorch/pull/115435

@htyu Triton already has it.

**Comment by manman-ren** [comment]
> Nice finding, @manman-ren ! I rerun the command in the summary with FMA manually disabled in triton, and the test can pass now. How can we properly fix this?

We need to figure out the llvm pass that caused this and fix it there.

**Comment by htyu** [comment]
> > Nice finding, @manman-ren ! I rerun the command in the summary with FMA manually disabled in triton, and the test can pass now. How can we properly fix this?
> 
> We need to figure out the llvm pass that caused this and fix it there.

It looks like FP fusion is on by default: https://github.com/openai/triton/blob/1eb0a31ffe3b1a0e5ea6cb298a7d938ed61c7799/third_party/nvidia/backend/compiler.py#L69

**Comment by Chillee** [comment]
It's also not clear to me whether this is a "real" accuracy error or just random changes that happen to trigger our accuracy test.

**Comment by manman-ren** [comment]
> It's also not clear to me whether this is a "real" accuracy error or just random changes that happen to trigger our accuracy test.

I have the same thought. The old and new ptx both have fma, the new ptx has 4 extra fmas. This is all triggered by the llir change due to the llvm passes. See attached old vs new.llir.
[old-llir.txt](https://github.com/pytorch/pytorch/files/14854917/old-llir.txt)
[new-llir.txt](https://github.com/pytorch/pytorch/files/14854918/new-llir.txt)

```
  %132 = select <4 x i1> %73, <4 x float> %130, <4 x float> <float -0.000000e+00, float -0.000000e+00, float -0.000000e+00, float -0.000000e+00>
  %133 = fadd <4 x float> %78, %132
vs
  %131 = fadd <4 x float> %78, %130
  %133 = select <4 x i1> %73, <4 x float> %131, <4 x float> %78
```

Because of this,  %130 = fmul <4 x float> %115, %129 happens right before fadd for new.llir, thus more fma.

**Comment by htyu** [comment]
If fma already exists in PTX, the program is sort of on fast-math path already? Wondering if the older version works because the test is not strong enough to catch those cases.

**Comment by shunting314** [comment]
> If fma already exists in PTX, the program is sort of on fast-math path already? Wondering if the older version works because the test is not strong enough to catch those cases.

@manman-ren / @htyu  I think this can be  verified a bit if we 
1. get the kernel with benchmark harness (enable TORCHINDUCTOR_BENCHMARK_KERNEL)
2. manually (we can not do that automatically ATM) write the equivalent pytorch code for the kernel. 
3. Compare if the kernel using old/new llvm can match the result from pytorch code.

Or alternatively, we can also compare if the results generated using old/new llvm are close enough with each other. But using pytorch eager as baseline sounds more solid.

**Comment by manman-ren** [comment]
For the accuracy regression, do we know how big is the actual regression?

The only thing with TORCHINDUCTOR_BENCHMARK_KERNEL is that I actually tried it with TORCHINDUCTOR_UNIQUE_KERNEL_NAMES=1, but the list of kernel names are different. So I can't find triton_red_fused__to_copy_native_layer_norm_16, but there are a few other ones with a different number at the end. Is there a way to figure out the name mapping with and without TORCHINDUCTOR_BENCHMARK_KERNEL.

**Comment by shunting314** [comment]
> For the accuracy regression, do we know how big is the actual regression?

I'm copying logs from the failed CI job.
```
2024-04-01T09:17:28.1578899Z cuda train mixer_b16_224                      
2024-04-01T09:18:00.1162456Z W0401 09:18:00.115000 140044580151936 torch/_logging/_internal.py:1016] [6/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
2024-04-01T09:18:33.0654350Z E0401 09:18:33.064000 140044580151936 torch/_dynamo/utils.py:1405] RMSE (res-fp64): 0.00147, (ref-fp64): 0.00015 and shape=torch.Size([384]). res.dtype: torch.float32, multiplier: 3.000000, tol: 0.010000
2024-04-01T09:18:33.0659637Z E0401 09:18:33.064000 140044580151936 torch/_dynamo/utils.py:1319] Accuracy failed for key name blocks.4.mlp_tokens.fc1.bias
2024-04-01T09:18:33.0675473Z fail_accuracy
```

At least for this job, it's on the boundary of passing the test. The formula is [here](https://github.com/pytorch/pytorch/blob/98e5238ad8b23d62f52549c88ee39f7cbacb46d8/torch/_dynamo/utils.py#L1403). But local run of the dashboard run may show different numerics. 


> So I can't find triton_red_fused__to_copy_native_layer_norm_16, but there are a few other ones with a different number at the end.

How many kernels are generated with name like `r'triton_red_fused__to_copy_native_layer_norm_[0-9]+'`? If not many, probably, we can try each of them

**Comment by manman-ren** [comment]
> How many kernels are generated with name like r'triton_red_fused__to_copy_native_layer_norm_[0-9]+'? If not many, probably, we can try each of them

There are 3 of them when enabling TORCHINDUCTOR_BENCHMARK_KERNEL, but 2 of them without. How can I find the wrapper py file for the 3 kernels?

**Comment by shunting314** [comment]
> There are 3 of them when enabling TORCHINDUCTOR_BENCHMARK_KERNEL, but 2 of them without. How can I find the wrapper py file for the 3 kernels?

The comment around the kernel definition in the wrapper should point you the path for each kernel.

**Comment by manman-ren** [comment]
I got the py file from the location info in the ttgir file. Attaching them:
[4-py.txt](https://github.com/pytorch/pytorch/files/14877904/4-py.txt)
[12-py.txt](https://github.com/pytorch/pytorch/files/14877905/12-py.txt)
